rmw_fastrtps_cpp/src/functions.cpp

```diff
@@ -1750,10 +1702,9 @@ extern "C"
             {
                 void *data = guard_conditions->guard_conditions[i];
                 GuardCondition *guard_condition = (GuardCondition*)data;
-                if(!guard_condition->getHasTriggered())
-                {
-                    guard_conditions->guard_conditions[i] = 0;
-                }
+                // Guard conditions persist (don't set to null)
+                // We call getHasTriggered to reset the trigger state.
```

Comment:
Just so we're on the same page, this is how I currently imagine things, maybe I am missing some part, so I'll describe it and perhaps you could correct me if I have it wrong? Currently you have to:

- Add guard conditions to the wait set once per executor (rather than once per call to rmw_wait)

And to preserve the ability to tell which guard condition was triggered you would need to:

- Add guard conditions to a list of guard conditions once per executor
- Add a pointer for each guard condition in the list once per call to rmw_wait

Is there other boilerplate that you need? If I'm right, you should no longer need to check which guard conditions were set to null each time in the executor (unless we add code to do debugging later). There is more work to be done in each rmw implementation (which we're talking about here), but I believe they are already setup to do that anyways.

The slight benefit to debugging, that I was thinking about, was being able to put a debug line in which printed a message, for example, when rmw_wait had been woken up by a node adding something new, rather than due to shutdown or ctrl-c. This would not be as convenient within the rmw implementations since they don't have the context about what each which guard condition they were given is used for, but the executor does.

Hopefully we can come to a consensus that this a valuable thing to preserve, rather than just giving in to the will of others. Honestly, if you think it's not a good idea then I'd said we do what you want since you're working on it actively at the moment.

As for the pattern of setting non-ready things to null, that's the pattern used by the `select` system call. My gut reaction is that it's probably the best solution because I've never seen a better one, but if you find one I'd love to hear about it. It is definitely an inconvenient interface, but I had written that off as one of the costs of doing this in C.